### PR TITLE
GEOMESA-1909,GEOMESA-1913,GEOMESA-1932,GEOMESA-2031 FSDS atomic writes and optional encoding

### PR DESCRIPTION
**Pull request by elahrvivaz**
* FileSystem API modified to remove schema management (handled by data store)
* FileSystem replaced with FileContext for atomic rename functionality
* Encoding is only required when creating a schema, otherwise it's read from metadata
* Configuration options must be passed in through 'fs.config'
* Partition schemes are SPI loaded and extensible
* Path status is cached to speed up file system ops
* MethodProfiling modified to support profiling Long values

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

File: build/cqs.tsv [hunk 1]
```diff
@@ -241,6 +241,7 @@ org.apache.hadoop:hadoop-annotations	2.6.4	provided
 org.apache.hadoop:hadoop-client	2.6.4	provided
 org.apache.hadoop:hadoop-common	2.6.4	provided
 org.apache.hadoop:hadoop-distcp	2.6.4	provided
+org.apache.hadoop:hadoop-hdfs	2.6.4	provided
```
Comment (elahrvivaz) [inline build/cqs.tsv]: covered by https://dev.eclipse.org/ipzilla/show_bug.cgi?id=8504

**Comment by elahrvivaz** [comment]
i've tested pretty thoroughly against local and hdfs, but need to test against s3 and through spark.

**Comment by elahrvivaz** [comment]
I changed the java API to use some Java 8 classes (i.e. `Optional`) - are we still trying to support java 7 as a runtime? I know we require java 8 for building now...

**Comment by jahhulbert-ccri** [comment]
@elahrvivaz no more java7 java 8 is minimum for geomesa :) esp since the java8 time updates.

**Comment by jnh5y** [comment]
@jahhulbert-ccri Emilio and I have assigned you to review and merge.

**Comment by jahhulbert-ccri** [comment]
Lets update this to remove the reference to "fs.encoding" -> parquet

http://www.geomesa.org/documentation/user/filesystem/sql.html

File: docs/user/filesystem/usage.rst [hunk 2]
```diff
@@ -23,11 +23,14 @@ FileSystem Data Store Parameters
...
-``fs.path *``       String The root path to write and read data from (e.g. s3a://mybucket/datastores/testds)
-``fs.encoding *``   String The encoding of the stored files. Provided implementations are "parquet" and "orc".
-``fs.read-threads`` Int    The number of threads used for queries
-``fs.config``       String Configuration values, formatted as a Java properties file, that will be passed to the underlying FileSystem
-=================== ====== ===========================================================================================================
+=============================== ====== ===================================================================================
+Parameter                       Type   Description
+=============================== ====== ===================================================================================
+``fs.path *``                   String The root path to write and read data from (e.g. s3a://mybucket/datastores/testds)
+``fs.encoding *``               String The encoding of the stored files. Provided implementations are "parquet" and "orc".
```
Comment (jahhulbert-ccri) [inline docs/user/filesystem/usage.rst]: should be optional now

Comment (elahrvivaz) [reply docs/user/filesystem/usage.rst]: thanks, i've updated the docs and removed all the unnecessary references to encoding (including in the spark sql example), but kept it there for ingest.

**Pull request merged**